Improvements to mortar friction-related code

recuero opened the pull request:
Small improvements. Also trying to figure out the reason of an intermittent (?) error in a 3d mortar test.

Discussed with @lindsayad.

https://civet.inl.gov/job/991101/

(#17495)

recuero committed:
Performance and clarity improvements to mortar friction-related code (#17495)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/991973/) on c5d5320 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20620/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/991981/) on c5d5320 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7c8de4</th>
      <th colspan=4>#20620 c5d532</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.53%</td>
      <td>82.53%</td>
      <td>+0.00%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73619</td>
      <td>73619</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15583</td>
      <td>15582</td>
      <td>-1</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20620/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20620/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>7c8de4</th>
      <th colspan=4>#20620 c5d532</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.07%</td>
      <td>89.06%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2925</td>
      <td>2924</td>
      <td>-1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>359</td>
      <td>359</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20620/coverage_diff/contact.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20620/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20620/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero committed:
Make MortarFrictionalPressureVectorAux compute at TIMESTEP_END

The test frictional-mortar-3d-status had been intermittently failing with components
 of the computed tangential vector that were far from the reference results. This appeared
to happen in parallel runs. The fact that the auxiliary kernel was not computing the vector
 at TIMESTEP_END did not guaranteed a converged solution. Since this aux kernel is there to help
with analysis, we can in principle just run it at TIMESTEP_END avoiding the performance hit and
generating always the converged aux kernel solution.

Credit to Alex Lindsay.

recuero committed:
Use superlu_dist for contact problems (#17495)

Alex and I have seen contact problems that run well fail with the
'PC failed due to FACTOR_OUTMEMORY' error intermittently. Switching to
superlu_dist as the sparse system solver appears to circumvent this issue. So
I am changing the settings of a number of mortar contact tests and hopefully
avoid these intermittent issues.

Credit to Alex Linday.

recuero commented:
@lindsayad, thanks for digging into this intermittent misbehavior!

lindsayad reviewed (changes requested):
`-pc_factor_mat_solver_package` has been deprecated since the PETSc 3.9.x series (see https://github.com/idaholab/moose/issues/11780) in favor of `-pc_factor_mat_solver_type`

recuero committed:
Replace -pc_factor_mat_solver_package with -pc_factor_mat_solver_type in contact module.

Address Alex's review.

lindsayad reviewed (approved):
Looks like a solid find-and-replace 😄 the indentation doesn't look as good, but it doesn't look bad either
